### ISSUE TITLE
Metadata: User stories for named channels

### ISSUE THREAD

**Issue by tischi** [ome/ngff#173]
I have the following user story:

Batch analysis of many multi-channel images, where the channel that contains the nuclei should be segmented.
The data is messy in a sense that the nucleus channel sometimes is the first one, and sometimes the second one.
To deal with this it seems useful to me to have something like `int channelIndex = omeZarrDataset.getChannelIndex( String channelID )`, where `channelID` in my case would be `"nuclei"`.  To be able to do this, I think some optional metadata inside the OME-Zarr that would allow one store and retrieve this mapping could be useful.

What do people think?

Does anyone else have a need for named channels?

ping @bogovicj @joshmoore

**Comment by d-v-b** [comment]
[Here's](http://cfconventions.org/Data/cf-conventions/cf-conventions-1.8/cf-conventions.html#_labels_and_alternative_coordinates) how this situation is handled in the climate sciences. Basically you associate the axis of the data with _another_ piece of data (in this case, an array of strings), which gives meaning to each element along the axis. This would actually be useful for non-discrete axes, too, and it would greatly simplify the specification of coordinate transformations (instead of defining a function that transforms one grid to another, you would just store the output grid directly)

Personally I think we should adopt as much of the climate science stuff as possible! I think their framework could solve a lot of our problems.

**Comment by ieivanov** [comment]
I do agree that metadata on channel names is critical for interpretation of a dataset. Spatial and temporal axes have units, which are encoded in the ome metadata: https://ngff.openmicroscopy.org/0.5/#axes-md. The channel dimension is discrete and should be associated with a label (i.e. channel name) at each index.

Our zarr stores often contain 7-8 channels (examples [here](https://www.ebi.ac.uk/biostudies/bioimages/studies/S-BIAD1287), I suggest looking at [figure_3.ome.zarr](https://ftp.ebi.ac.uk/biostudies/fire/S-BIAD/287/S-BIAD1287/Files/mantis/figure_3.ome.zarr.zip) and [figure_1.ome.zarr](https://ftp.ebi.ac.uk/biostudies/fire/S-BIAD/287/S-BIAD1287/Files/mantis/figure_1.ome.zarr.zip); sorry they had to be zipped,  I can share just the metadata here if that's useful), and we fetch the right channel for downstream analysis by its name. We store information on the channel names in the omero metadata: https://ngff.openmicroscopy.org/0.5/#omero-md. This field is labeled as transitional - I think it's important for the community to identify a new strategy for (1) storing channel names in the ome metadata and (2) storing information on how a channel should be rendered by a viewer.

I like @d-v-b's suggestion of adopting ideas from other fields. Happy to discuss further if helpful.

**Comment by jluethi** [comment]
As @lubianat referred to this from the summary issue, I just wanted to briefly +1 the use case for named channels here! 

Our user story: When dealing with image analysis of OME-Zarrs, we mostly cared about "identifiers that users can provide to select a channel". We've used either the channel index (very unintuitive for users), the channel label (as found in omero metadata, very useful for users if it is set it correctly) or additional metadata (e.g. we added a wavelength_id field to many of our omero metadata fields to group different label names that actually are the same wavelength in multiplexed imaging). For that, we heavily relied on (customized) omero metadata for channel selection in the processing space.

I'd be very interested in broader, shared definitions of how one can refer to a given channel besides its index, like what we've treated the transitional omero metadata field so far.

**Comment by lubianat** [comment]
I am struggling to find _where_ I had seen it, but I've seen some implementation using the _color_ of the channel in `omero.channels` as an unique identifier. Probably a bad idea, but a workaround needed due to the lack of named channels. 

Edit: found it!! (via https://github.com/ome/ngff-spec/pull/87

related to the redo of omero fields, here is an insteesting scenario where color is being used as a local ID for a channel

https://github.com/JurgenKriel/macro_candi

_Channels are resolved by colour, not index. canmac/io/channels.py reads the OME-Zarr omero metadata and matches hex colours, so channel order changes don't silently mis-assign data. Edit COLORS there for your fluorophores._

<img width="927" height="89" alt="Image" src="https://github.com/user-attachments/assets/c560093d-69ae-47c3-95d5-51f11dd9b25c" />

**Comment by jo-mueller** [comment]
Some 2or3cts from my side:

The first idea that came to my mind to make this a bit nicer, was to add the channels metadata directly where the channels are described, i.e. in the axes metadata, which would look something like this:

```json
{
  "name": "physical",
  "axes": [
    {
      "name": "c",
      "type": "channel",
      "channels": [
        {
          "name": "DAPI",
          "type": "fluorescent",
          "color": "#808080",
          ...
          }
      ]
    },
    {
      "name": "z",
      "type": "space",
    }
  ]
}
```

which would be neat in the sense that the channel information sits directly adjacent to where the existence of a channel dimension is declared in the first place.

**But**:
- images need not necessarily have a channel dimension (brightfield is single-channel with no extra channel dimension) but one would still want to declare the channel *somewhere*. The channel axis doesn't exist in this case, so that's not an option.
- The other argument against plugging channel info into the axis metadata comes from a coordinate transform point of view: If you have two coordinate systems with a simple translation in-between, would you need to declare the full channel metadata in both? Duplication is undesirable, so one would declare the channel metadata elsewhere and reference it here.

Which brings us back to the status quo: The channel information is fundamentally tied to the image in a semantic sense, and not subject to reordering from a coordinate transform. Maybe we'd need to say this somewhere: coordinate transformations are not defined for dimensions of `"discrete": true` and transforms MUST fall back to default values for such dimensions (i.e. 1.0 for scale, 0.0 for translate, etc) This would make sure that transforms don't reorder the channel dimension or do other weird things along this axis.

So...I think this brings us to the path of replacing the omero metadata by a proper `"channels"` attribute, which should sit under the `multiscales` metadata (or in the `attributes` of an `MultiscaleNode` in a post-RFC8 world). There we could collect from the `omero` metadata what feels useful and enrich it with what's missing? I.e., wavelengths, dye, name, color and brightness, etc.

**Comment by d-v-b** [comment]
I highly recommend making the decisions about how to logically organize this information by thinking about the different consumers of it. 
- Labels for a discrete axis (like channels, but also many other things) would be consumed by any application that wants to faithfully represent the basic structure of the data, and support label-based lookup of data values (e.g., a software that allows the user to index data using axis labels instead of plain integers). Software at all layers of the stack -- from CLIs to rich GUIs -- could use this.
- Instructions for rendering a particular view of the data (e.g., the `color` field above). First, this is not channel-specific: folks might use these instructions for segmentation data, or any other data with a discrete axis. Second, I would argue that "how to display this on a screen" narrowly relevant to a small subset of (very important) consumers, namely image visualization tools, and even then I worry that different visualization tools have different models for this kind of information. Some tools take a color string, but other tools like neuroglancer are much more expressive and can take raw shader code as a definition of how data should be rendered.

Another important point about rendering settings: it's definitely important to store metadata about how visualization tools should display data. But I don't think this is an attribute of a single image. It's an attribute of _at least one view_ of an image. That distinction matters because people often combine separate images (like combining multiple segmentations over the raw data). For openorganelle, information about rendering was stored at the image collection level as an array of "views". We needed this because we found that even a single image might need different contrast limits in different regions, and when overlaying multiple images, we might use differnent colors per image depending on the number of images in the view. If two images (or two channels) claim the color blue, then you simply can't overlay them. So we solved this by picking rendering settings based on everything we knew about a view -- which images, how far zoomed in, etc. If ome-zarr approaches this by binding single rendering settings to entire images, then it will create a lot of complexity for folks trying to create multi-image scenes. I would avoid this if possible.

**Comment by will-moore** [comment]
Yes, even in OMERO we store multiple rendering settings per Image. Users only see 1 per user, but it's possible to store multiple settings per user.
I think there's a distinction between *what* the channel is (aquisition metadata) and how to view it (rendering settings)

**Comment by d-v-b** [comment]
> I think there's a distinction between _what_ the channel is (aquisition metadata) and how to view it (rendering settings)

💯 and these have distinct (but overlapping) consumers.

**Comment by lubianat** [comment]
(hopefully it is okay; I moved the last comments to #580 to try and center scoping/design discussions there, keeping this issue for user stories only)

**Comment by jo-mueller** [comment]
EDIT: moved to #580

**Comment by lubianat** [comment]
@ieivanov it has been a while, but I'll take your offer of sharing just the metadata for your user story :) 
the zip files are quite big!

**Comment by ieivanov** [comment]
Sorry for the delay, here are a couple of examples:

[figure_1.zarr/0/0/0/.zattrs](https://github.com/user-attachments/files/31275824/zattrs.json)

[figure_3.zarr/0/DCP1B/000000/.zattrs](https://github.com/user-attachments/files/31275883/zattrs.json)

One issue I foresee if indexing by color - we do a lot of label-free imaging and frequently have several grayscale channels (for example `Phase`, `Retardance`, `Orientation`, `Brightfield`, `State0` through `State4`). We also frequently predict images with AI models - here `Nuclei-prediction` and `Membrane-prediction`, and save them along with other label-free or fluorescent channels. These "virtually stained" channels are assigned a color for visualization purposes, but they are not inherently linked to a fluorophore emission or excitation wavelength.
